File: client-extensions/swift/Verified.swift
import Foundation
import CryptoKit
public enum Verified {
    public static func readFile(client: AgentTrunk, workspace: String, key: String, revision: String, path: String) async throws -> Data {
        guard revision.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil,
              !path.isEmpty, path.utf8.count <= 512, !path.contains("\\"),
              !path.unicodeScalars.contains(where: { $0.value < 32 || $0.value == 127 }),
              !path.components(separatedBy: "/").contains(where: { ["", ".", ".."].contains($0) }) else { throw AgentTrunkError.invalidResponse }
        let pin = try await client.contexts.inspect(trunkId: workspace, contextKey: key, ref: revision)
        guard pin.revision.id == revision, let file = pin.revision.files.first(where: { $0.path == path }),
              (0...1_000_000).contains(file.size) else { throw AgentTrunkError.invalidResponse }
        let data = try await client.contexts.readFile(trunkId: workspace, contextKey: key, resourcePath: path, ref: revision)
        let hash = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        guard data.count == file.size, hash == file.sha256 else { throw AgentTrunkError.invalidResponse }
        return data
    }
}

File: sdk/rust/src/verified.rs
use crate::prelude::*;
use sha2::{Digest, Sha256};
fn digest(s: &str) -> bool { s.len()==64 && s.bytes().all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b)) }
pub async fn read_file(client: &AgentTrunk, workspace: &str, key: &str, revision: &str, path: &str) -> Result<Vec<u8>, ApiError> {
    if !digest(revision) || path.is_empty() || path.len()>512 || path.chars().any(|c| c=='\\' || c.is_control()) || path.split('/').any(|s|s.is_empty()||s=="."||s=="..") {return Err(ApiError::InvalidHeader);}
    let pin=client.contexts.inspect(workspace,key,&InspectQueryRequest{r#ref:Some(revision.into())},None).await?;
    if pin.revision.id!=revision {return Err(ApiError::InvalidHeader);}
    let file=pin.revision.files.iter().find(|f|f.path==path).ok_or(ApiError::InvalidHeader)?;
    if file.size<0 || file.size>1_000_000 || !digest(&file.sha256){return Err(ApiError::InvalidHeader);}
    let data=client.contexts.read_file(workspace,key,path,&ReadFileQueryRequest{r#ref:revision.into()},None).await?.collect().await?;
    if data.len() as i64 != file.size || format!("{:x}",Sha256::digest(&data))!=file.sha256 {return Err(ApiError::InvalidHeader);}
    Ok(data)
}
